### PR TITLE
feat: Add license checking tool

### PR DESCRIPTION
**Pull request by srebhan**
- [ ] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

This PR adds a small tool to verify the license information in `docs/LICENSE_OF_DEPENDENCIES.md` against the repository content of the individual dependencies.

To run the tool first build it with `make build_tools` and then 
```shell
./tools/license_checker/license_checker -whitelist ./tools/license_checker/data/whitelist -cache .licenses.cache .
```
This will use the whitelisting for `github.com/eclipse/paho.mqtt.golang` and caches the classification results.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -169,6 +169,8 @@ require (
 	modernc.org/sqlite v1.17.3
 )
 
+require github.com/google/licensecheck v0.3.1
```
Comment (powersj) [inline go.mod:172]: can you move this up to the require that lists all our direct dependencies so we don't have yet another require section?

File: go.mod [hunk 2]
```diff
@@ -362,7 +364,7 @@ require (
 	go.uber.org/multierr v1.6.0 // indirect
 	golang.org/x/crypto v0.0.0-20220315160706-3147a52a75dd // indirect
 	golang.org/x/exp v0.0.0-20200513190911-00229845015e // indirect
-	golang.org/x/mod v0.5.1 // indirect
+	golang.org/x/mod v0.5.1
```
Comment (powersj) [inline go.mod:367]: same here

File: tools/license_checker/cache.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+func (c *cache) Get(pkg *packageInfo) (string, float64, bool) {
+	entry, found := c.Entries[pkg.name+"@"+pkg.version]
+	if !found {
+		return "", 0.0, false
+	}
+	now := time.Now()
+	deadline := entry.Update.Add(c.Expiry)
+	if c.Expiry > 0 && now.After(deadline) {
+
+		return entry.Spdx, entry.Confidence, false
```
Comment (powersj) [inline tools/license_checker/cache.go]:
```suggestion
	if c.Expiry > 0 && now.After(deadline) {
		return entry.Spdx, entry.Confidence, false
```

**Review by powersj — COMMENTED** [review]
Briefly looked through this, this needs a README on how to run it and details on the whitelist documented.

Other than the items in the whitelist, have you run into many mistakes or issues with trying to parse the licenses? I see the threshold setting and was wondering if you had to play with it to get this to work consistently?

**Comment by srebhan** [comment]
@MyaLongmire and @sspaink removed the cache as discussed and added to nightly build. Please have a look and let me know what you think!

File: .circleci/config.yml [hunk 4]
```diff
@@ -232,6 +232,25 @@ jobs:
...
+      - run: 'sh ./tools/license_checker/license_checker -whitelist ./tools/license_checker/data/whitelist'
+      - save_cache:
+          name: 'go module cache'
+          key: go-mod-v1-{{ checksum "go.sum" }}
+          paths:
+            - '/go/pkg/mod'
+      - persist_to_workspace:
+          root: '/go'
+          paths:
+            - '*'
```
Comment (sspaink) [inline .circleci/config.yml:252]: Not sure if this is necessary, I assume nothing was added to the `/go` dir that would need to be saved.

Comment (srebhan) [reply .circleci/config.yml]: Removed. Thanks for catching this!

**Review by sspaink — APPROVED** [review]
Thank you for making this!

**Review by powersj — APPROVED** [review]
Awesome! Thank you!

**Pull request merged**